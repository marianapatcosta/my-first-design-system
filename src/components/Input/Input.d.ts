import React, { ChangeEvent, CSSProperties, MouseEventHandler } from 'react';
declare type InputType = 'button' | 'checkbox' | 'color' | 'date' | 'datetime-local' | 'email' | 'file' | 'hidden' | 'image' | 'month' | 'number' | 'password' | 'radio' | 'range' | 'reset' | 'search' | 'submit' | 'tel' | 'text' | 'time' | 'url' | 'week' | undefined;
export interface InputProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * Input label
     */
    label?: string;
    /**
     * Input name
     */
    name?: string;
    /**
     * Input value
     */
    value: string;
    /**
     * Input type
     */
    type?: InputType;
    /**
     * Input's displaying text when no option is selected
     */
    placeholder?: string;
    /**
     * Is Input disabled?
     */
    disabled?: boolean;
    /**
     * what is maxLength allowed for the input field?
     */
    maxLength?: number;
    /**
     * is autoComplete functionality available?
     */
    autoComplete?: string;
    /**
     * is autoFocus functionality available?
     */
    autoFocus?: boolean;
    /**
     * text to display in case of error
     */
    errorMessage?: string;
    /**
     * icon to display in input field
     */
    icon?: string;
    /**
     * text to describe the input icon (for accessibility purposes)
     */
    iconText?: string;
    /**
     * Change handler
     */
    onChange: (event: ChangeEvent<HTMLInputElement>) => void;
    /**
     * Blur handler
     */
    onBlur?: (event: ChangeEvent<HTMLInputElement>) => void;
    /**
     * Click handler on icon click
     */
    onIconClick?: MouseEventHandler<HTMLImageElement>;
}
export declare const Input: React.FC<InputProps>;
export {};
