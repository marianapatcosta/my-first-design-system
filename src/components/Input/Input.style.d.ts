export declare const StyledInputWrapper: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledInputField: import("styled-components").StyledComponent<"div", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledInput: import("styled-components").StyledComponent<"input", any, {
    errorMessage?: string | undefined;
}, never>;
export declare const StyledLabel: import("styled-components").StyledComponent<"label", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledInputIcon: import("styled-components").StyledComponent<"img", any, {}, never>;
export declare const StyledInputError: import("styled-components").StyledComponent<"p", any, {}, never>;
