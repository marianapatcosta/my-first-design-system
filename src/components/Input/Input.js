var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyledInputWrapper, StyledInput, StyledInputField, StyledLabel, StyledInputIcon, StyledInputError, } from './Input.style';
export var Input = function (_a) {
    var id = _a.id, className = _a.className, label = _a.label, disabled = _a.disabled, placeholder = _a.placeholder, errorMessage = _a.errorMessage, icon = _a.icon, iconText = _a.iconText, onIconClick = _a.onIconClick, props = __rest(_a, ["id", "className", "label", "disabled", "placeholder", "errorMessage", "icon", "iconText", "onIconClick"]);
    return (_jsxs(StyledInputWrapper, { children: [!!label && (_jsx(StyledLabel, __assign({ htmlFor: id, disabled: disabled }, { children: label }), void 0)),
            _jsxs(StyledInputField, __assign({ className: className, disabled: disabled }, { children: [_jsx(StyledInput, __assign({ "data-testid": 'input-field', "aria-label": !!label ? label : placeholder, id: id, disabled: disabled, errorMessage: errorMessage, placeholder: placeholder }, props), void 0),
                    icon && (_jsx(StyledInputIcon, { "data-testid": 'input-icon', "aria-label": !!onIconClick ? 'click me' : 'image', alt: iconText, src: icon, onClick: onIconClick, tabIndex: onIconClick ? 0 : -1 }, void 0))] }), void 0),
            !!errorMessage && _jsx(StyledInputError, { children: errorMessage }, void 0)] }, void 0));
};
Input.defaultProps = {
    id: '',
    className: '',
    style: {},
    label: '',
    name: '',
    disabled: false,
    maxLength: 150,
    type: 'text',
    autoComplete: 'off',
    autoFocus: false,
    value: '',
    placeholder: '',
    errorMessage: '',
    icon: '',
    iconText: '',
    onChange: function () { return null; },
};
