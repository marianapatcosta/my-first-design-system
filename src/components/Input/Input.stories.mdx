import { Meta, Story, ArgsTable, Canvas, Preview } from '@storybook/addon-docs/blocks';
import { Input } from "./Input";
import { Success } from '../../assets/icons';

<Meta title="Design System/Input" component={Input} />

# Input
<br/><br/> 

### Displays a customized Input field, to insert small amount of data

Use input field when it is supposed that the user provides some written information

<Preview withToolbar>
  <Story name="Standard">
    <Input value="hello" />
  </Story>
</Preview>

<ArgsTable of={Input} />

### Input with Label:

  <Story name="WithLabel">
    <Input value="hello" label="Input" />
  </Story>

### Input with Icon:

  <Story name="WithIcon">
    <Input value="hello" icon={Success} />
  </Story>

### Input with Error:
  <Story name="WithError">
    <Input value="hello" errorMessage="This is an error!" />
  </Story>


### Playground

Experiment with Input Component with the Controls addon in the Canvas tab.

export const Template = (args) => <Input {...args} />

<Canvas>
  <Story name="Controls" args={{
    value: 'hello'
  }}>
    {Template.bind({})}
  </Story>
</Canvas> 
