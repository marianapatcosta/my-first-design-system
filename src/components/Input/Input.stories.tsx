import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Input, InputProps } from './Input';
import { Tick } from '../../assets/icons';

export default {
  title: 'Input',
  component: Input,
  args: { value: 'hello' },
} as Meta;

const Template: Story<InputProps> = (args) => <Input {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const WithLabel = Template.bind({});
WithLabel.args = { label: 'Input' };

export const WithIcon = Template.bind({});
WithIcon.args = { icon: Tick };

export const WithError = Template.bind({});
WithError.args = { errorMessage: 'This is an error!' };
