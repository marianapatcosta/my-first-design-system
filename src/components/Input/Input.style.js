var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, disabledStyles, colors, typography } from '../../styles';
export var StyledInputWrapper = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  margin: 1rem 0;\n  width: 100%;\n  max-width: 15rem;\n"], ["\n  ", "\n  margin: 1rem 0;\n  width: 100%;\n  max-width: 15rem;\n"])), baseStyles);
export var StyledInputField = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  position: relative;\n  ", ";\n"], ["\n  position: relative;\n  ", ";\n"])), function (_a) {
    var disabled = _a.disabled;
    return disabled && " " + disabledStyles;
});
export var StyledInput = styled.input(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", "\n  display: block;\n  color: ", ";\n  font-family: ", ";\n  font-size: 95%;\n  width: 100%;\n  border: 0.0625rem solid ", ";\n  background-color: ", ";\n  box-shadow: 0 0.063rem 0.25rem ", ";\n  box-sizing: border-box;\n  padding: 0.3rem 2rem 0.3rem 0.625rem;\n  height: 2rem;\n\n  :-webkit-autofill,\n  :-webkit-autofill:hover,\n  :-webkit-autofill:focus,\n  :-webkit-autofill:active {\n    -webkit-text-fill-color: ", ";\n    -webkit-box-shadow: 0 0 0 2rem ", " inset !important;\n  }\n\n  ", "\n"], ["\n  ", "\n  display: block;\n  color: ", ";\n  font-family: ", ";\n  font-size: 95%;\n  width: 100%;\n  border: 0.0625rem solid ", ";\n  background-color: ", ";\n  box-shadow: 0 0.063rem 0.25rem ", ";\n  box-sizing: border-box;\n  padding: 0.3rem 2rem 0.3rem 0.625rem;\n  height: 2rem;\n\n  :-webkit-autofill,\n  :-webkit-autofill:hover,\n  :-webkit-autofill:focus,\n  :-webkit-autofill:active {\n    -webkit-text-fill-color: ", ";\n    -webkit-box-shadow: 0 0 0 2rem ", " inset !important;\n  }\n\n  ",
    "\n"])), baseStyles, colors.font, typography.type.body, colors.transparent, colors.primary, colors.shadow, colors.font, colors.primary, function (_a) {
    var errorMessage = _a.errorMessage;
    return !!errorMessage &&
        "border-color: " + colors.red + ";\n    box-shadow: none;\n    ";
});
export var StyledLabel = styled.label(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  display: block;\n  color: ", ";\n  margin-bottom: 0.3rem;\n  margin-left: 0.3rem;\n  ", "\n"], ["\n  display: block;\n  color: ", ";\n  margin-bottom: 0.3rem;\n  margin-left: 0.3rem;\n  ",
    "\n"])), colors.highlight, function (_a) {
    var disabled = _a.disabled;
    return disabled && disabledStyles + "; color: " + colors.disabled;
});
export var StyledInputIcon = styled.img(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  position: absolute;\n  width: 1rem;\n  height: 1rem;\n  top: 0;\n  bottom: 0;\n  margin: auto 0;\n  right: 0.625rem;\n  ", ";\n"], ["\n  position: absolute;\n  width: 1rem;\n  height: 1rem;\n  top: 0;\n  bottom: 0;\n  margin: auto 0;\n  right: 0.625rem;\n  ", ";\n"])), colors.icon);
export var StyledInputError = styled.p(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  color: ", ";\n  font-size: 95%;\n  margin-top: 0.3rem;\n"], ["\n  color: ", ";\n  font-size: 95%;\n  margin-top: 0.3rem;\n"])), colors.red);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
