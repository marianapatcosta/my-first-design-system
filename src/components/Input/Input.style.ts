import styled from 'styled-components';
import { baseStyles, disabledStyles, colors, typography } from '../../styles';

export const StyledInputWrapper = styled.div`
  ${baseStyles}
  margin: 1rem 0;
  width: 100%;
  max-width: 15rem;
`;

export const StyledInputField = styled.div<{ disabled?: boolean }>`
  position: relative;
  ${({ disabled }) => disabled && ` ${disabledStyles}`};
`;

export const StyledInput = styled.input<{ errorMessage?: string }>`
  ${baseStyles}
  display: block;
  color: ${colors.font};
  font-family: ${typography.type.body};
  font-size: 95%;
  width: 100%;
  border: 0.0625rem solid ${colors.transparent};
  background-color: ${colors.primary};
  box-shadow: 0 0.063rem 0.25rem ${colors.shadow};
  box-sizing: border-box;
  padding: 0.3rem 2rem 0.3rem 0.625rem;
  height: 2rem;

  :-webkit-autofill,
  :-webkit-autofill:hover,
  :-webkit-autofill:focus,
  :-webkit-autofill:active {
    -webkit-text-fill-color: ${colors.font};
    -webkit-box-shadow: 0 0 0 2rem ${colors.primary} inset !important;
  }

  ${({ errorMessage }) =>
    !!errorMessage &&
    `border-color: ${colors.red};
    box-shadow: none;
    `}
`;

export const StyledLabel = styled.label<{ disabled?: boolean }>`
  display: block;
  color: ${colors.highlight};
  margin-bottom: 0.3rem;
  margin-left: 0.3rem;
  ${({ disabled }) =>
    disabled && `${disabledStyles}; color: ${colors.disabled}`}
`;

export const StyledInputIcon = styled.img`
  position: absolute;
  width: 1rem;
  height: 1rem;
  top: 0;
  bottom: 0;
  margin: auto 0;
  right: 0.625rem;
  ${colors.icon};
`;

export const StyledInputError = styled.p`
  color: ${colors.red};
  font-size: 95%;
  margin-top: 0.3rem;
`;
