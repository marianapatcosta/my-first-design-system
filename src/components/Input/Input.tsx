import React, { ChangeEvent, CSSProperties, MouseEventHandler } from 'react';
import {
  StyledInputWrapper,
  StyledInput,
  StyledInputField,
  StyledLabel,
  StyledInputIcon,
  StyledInputError,
} from './Input.style';

type InputType =
  | 'button'
  | 'checkbox'
  | 'color'
  | 'date'
  | 'datetime-local'
  | 'email'
  | 'file'
  | 'hidden'
  | 'image'
  | 'month'
  | 'number'
  | 'password'
  | 'radio'
  | 'range'
  | 'reset'
  | 'search'
  | 'submit'
  | 'tel'
  | 'text'
  | 'time'
  | 'url'
  | 'week'
  | undefined;

export interface InputProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * Input label
   */
  label?: string;
  /**
   * Input name
   */
  name?: string;
  /**
   * Input value
   */
  value: string;
  /**
   * Input type
   */
  type?: InputType;
  /**
   * Input's displaying text when no option is selected
   */
  placeholder?: string;
  /**
   * Is Input disabled?
   */
  disabled?: boolean;
  /**
   * what is maxLength allowed for the input field?
   */
  maxLength?: number;
  /**
   * is autoComplete functionality available?
   */
  autoComplete?: string;
  /**
   * is autoFocus functionality available?
   */
  autoFocus?: boolean;
  /**
   * text to display in case of error
   */
  errorMessage?: string;
  /**
   * icon to display in input field
   */
  icon?: string;
  /**
   * text to describe the input icon (for accessibility purposes)
   */
  iconText?: string;
  /**
   * Change handler
   */
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  /**
   * Blur handler
   */
  onBlur?: (event: ChangeEvent<HTMLInputElement>) => void;
  /**
   * Click handler on icon click
   */
  onIconClick?: MouseEventHandler<HTMLImageElement>;
}

export const Input: React.FC<InputProps> = ({
  id,
  className,
  label,
  disabled,
  placeholder,
  errorMessage,
  icon,
  iconText,
  onIconClick,
  ...props
}) => {
  return (
    <StyledInputWrapper>
      {!!label && (
        <StyledLabel htmlFor={id} disabled={disabled}>
          {label}
        </StyledLabel>
      )}
      <StyledInputField className={className} disabled={disabled}>
        <StyledInput
          data-testid='input-field'
          aria-label={!!label ? label : placeholder}
          id={id}
          disabled={disabled}
          errorMessage={errorMessage}
          placeholder={placeholder}
          {...props}
        />
        {icon && (
          <StyledInputIcon
            data-testid='input-icon'
            aria-label={!!onIconClick ? 'click me' : 'image'}
            alt={iconText}
            src={icon}
            onClick={onIconClick}
            tabIndex={onIconClick ? 0 : -1}
          />
        )}
      </StyledInputField>
      {!!errorMessage && <StyledInputError>{errorMessage}</StyledInputError>}
    </StyledInputWrapper>
  );
};

Input.defaultProps = {
  id: '',
  className: '',
  style: {},
  label: '',
  name: '',
  disabled: false,
  maxLength: 150,
  type: 'text',
  autoComplete: 'off',
  autoFocus: false,
  value: '',
  placeholder: '',
  errorMessage: '',
  icon: '',
  iconText: '',
  onChange: () => null,
};
