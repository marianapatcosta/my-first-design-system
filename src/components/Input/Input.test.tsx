import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Input } from './Input';
import { Tick as Success } from '../../assets/icons';

describe('Input', () => {
  const defaultProps = { label: 'Input', value: 'Hello', onChange: () => null };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Input {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should handle change events', () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <Input {...defaultProps} onChange={onChangeMock} />
    );
    const input = getByTestId('input-field') as HTMLInputElement;
    fireEvent.click(input, { target: { value: 'bye' } });
    expect(input.value).toBe('bye');
  });

  test('should handle blur events', () => {
    const onBlurMock = jest.fn();
    const { getByTestId } = render(
      <Input {...defaultProps} onBlur={onBlurMock} />
    );
    const input = getByTestId('input-field') as HTMLInputElement;
    fireEvent.blur(input);
    expect(onBlurMock).toBeCalled();
  });

  test('should return null on change if onChange prop is not passed', () => {
    const result =
      Input.defaultProps?.onChange && Input.defaultProps.onChange();
    expect(result).toBeNull();
  });

  test('should return null on blur if onChange prop is not passed', () => {
    const result = Input.defaultProps?.onBlur && Input.defaultProps.onBlur();
    expect(result).toBeNull();
  });

  describe('Disabled Input', () => {
    test('should not handle change events', () => {
      const onChangeMock = jest.fn();
      const {
        container: { firstChild },
      } = render(
        <Input {...defaultProps} disabled={true} onChange={onChangeMock} />
      );

      firstChild && fireEvent.click(firstChild);

      expect(onChangeMock).not.toBeCalled();
    });
  });

  describe('Input with icon', () => {
    test('should render icon if icon prop is passed', () => {
      const { getByTestId } = render(
        <Input {...defaultProps} icon={Success} />
      );

      const icon = getByTestId('input-icon');
      expect(icon).toBeTruthy();
      expect(icon).toHaveAttribute('src', Success);
    });

    test('should handle click events', () => {
      const onIconClickMock = jest.fn();
      const { getByTestId } = render(
        <Input {...defaultProps} icon={Success} onIconClick={onIconClickMock} />
      );
      const icon = getByTestId('input-icon');
      icon && fireEvent.click(icon);

      expect(onIconClickMock).toBeCalled();
    });
  });

  describe('Input with error', () => {
    test('should render error message', () => {
      const { getByText } = render(
        <Input {...defaultProps} errorMessage='There is an error' />
      );

      const input = getByText(/There is an error/i);
      expect(input).toBeTruthy();
    });
  });
});
