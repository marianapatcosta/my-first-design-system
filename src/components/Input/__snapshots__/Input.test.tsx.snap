// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders without errors and matches snapshot 1`] = `
<div
  class="sc-bdfBwQ hraWbH"
>
  <label
    class="sc-hKgILt jatIhT"
    for=""
  >
    Input
  </label>
  <div
    class="sc-gsTCUz ExCDi"
  >
    <input
      autocomplete="off"
      class="sc-dlfnbm jrCcpg"
      data-testid="input-field"
      id=""
      maxlength="150"
      name=""
      placeholder=""
      type="text"
      value="Hello"
    />
    
  </div>
</div>
`;
